### PR TITLE
Kitsune2 Bootstrap Server--Core

### PR DESCRIPTION
**Pull request by neonphog**
This is the core functionality of the kitsune2 bootstrap server code.

To make the PRs more manageable, I've broken them out into some preparatory work before this, and all the endpoint test after this.

- See the previous [Prep #9](https://github.com/holochain/kitsune2/pull/9) PR for that prep work.
- See the follow on [Testing #11](https://github.com/holochain/kitsune2/pull/11) PR for api endpoint tests and general notes about this effort in the PR description.

### CODE REVIEW

File: crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:19]: Do we want to move these into issues in the kitsune2 repo?

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:301]: So the idea is that if reading gets interrupted for some reason, it retries? What conditions would cause it to get interrupted? Should we limit the number of retries to avoid something like a disk failure causing infinite looping?

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:233]: Maybe we should move this value to a const in the config file.

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:226]: Maybe we should move this value to a const in the config file.

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:90]: I read this and thought it was a typo on "main_worker" -- maybe spelling it fully would be cleaer.

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:167]: It looks like there's only one error where the error string is more than just a "type" label (line 259 `                std::io::Error::other(format!("InvalidSignature: {err:?}"))`.

If we want to include any message along with a "type" label, then I'd want them split them out into 2 fields for clients to easily match on.

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:165]: Nitpick but some of these errors would better fit the 400 response type.

File: crates/bootstrap_srv/src/server.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/server.rs:259]: The error mentioned above.

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/space.rs:8]: Is there a reason to use `bytes::Bytes` instead of the wrapper type you made `SpaceId`?

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/space.rs:18]: same here - should this be a `SpaceId`?

File: crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:19]: This PR is against the kitsune2 repo. Sorry, not sure what you mean here.

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:313]: Clients only optionally send this header, and then (while technically required by the spec) may not always send the same byte count... so I thought best to guard on the actual data sent.

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:301]: This is the fundamental behavior of the core libc read function, carried forward from back in the days when interrupt requests were how processes were parallelized. When you call `write_all`, under the hood it is doing this exact loop check. There is no need to diverge from cannon here.

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/space.rs:89]: I don't understand the difference between this behavior and RwLock, but as long as others do that's fine with me.

File: crates/bootstrap_srv/src/config.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/config.rs:18]: mmm... I'm doing that as a DRY tool, since the same cleanup logic was required. Many other threads will be created by the tiny_http library itself, and who knows about any others for tracing libraries, etc. I'm not sure we need to enumerate them all.

File: crates/bootstrap_srv/src/space.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs:8]: Yes. I see this server / protocol being useful for WAN discovery outside of projects using kitsune itself. Perhaps it would actually be better to exist outside this monorepo, but that is more overhead for us. At the very least, I didn't want this to have any dependencies on other kitsune crates.

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:167]: @mattyg - For this and https://github.com/holochain/kitsune2/pull/10#discussion_r1854225991 (which I don't consider a nitpick, I agree these are errors in user input, not in what the server is doing), I wonder if the error reporting types, since they are defined in json, and statuses actually belong in the spec itself. Would you be okay with updating the spec and fixing these responses being a follow-on effort after these are merged?

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:165]: See https://github.com/holochain/kitsune2/pull/10#discussion_r1854281943

File: crates/bootstrap_srv/src/store.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/store.rs:9]:
```suggestion
//!   than would reasonably fit in RAM.
//! - We want to avoid requiring on a database server as an additional system dependency for ease of deployment and reduced overhead.
```

File: crates/bootstrap_srv/src/store.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/store.rs:13]: I'm confused by this line -- it sounds like the max number of tempfiles is the worker_thread_count?

Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:13]: Yes exactly. Since we're not using async code, the items from the pool can only be checked out by an active thread, and then they are returned before that thread can check out another. So the worker thread count is the *implicit* max count of active temp files that can be checked out from the pool.

There is a nuance here, however: when the tempfiles roll over from size, they are not re-added to the pool. So the pool will never grow beyond that, but while there are still read refs held for these other items, they will remain in memory.

File: crates/bootstrap_srv/src/store.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/store.rs:193]: I assume we want a lot testing more here, but i assume you know that :)

Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:193]: @mattyg - I'm considering the testing in [Testing #11](https://github.com/holochain/kitsune2/pull/11) as sufficient for covering this. If you don't agree, please let me know what else you'd like to see here before approving : )

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [inline crates/bootstrap_srv/src/space.rs:212]: would it be clearer to have an explicit "strategy" defined somewhere and this function to call it?

Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs:212]: For sure eventually this will need to change. But I'm not sure the scope of the change that will be required until we go to actually design the new strategy... so I'm inclined to keep this simple until needed.

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [reply crates/bootstrap_srv/src/space.rs:8]: In that case do you want to define a SpaceId or similar wrapper type in this crate just for readability?

File: crates/bootstrap_srv/src/space.rs
Comment (mattyg) [reply crates/bootstrap_srv/src/space.rs:89]: I see so we need the write lock for a long time because the re-write code is processing and mutating the data in one go, rather than constructing a new list with only a read lock, then at the end getting the write lock and writing it all at once.

And there is no downside to allowing readers at the same time because a single entry will be mutated atomically, and so there's no risk of a reader getting some partially-written entry change? (forgive my ignorance)

File: crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:19]: [Done](https://github.com/holochain/kitsune2/issues)

File: crates/bootstrap_srv/src/store.rs
Comment (mattyg) [reply crates/bootstrap_srv/src/store.rs:193]: whoops my bad

File: crates/bootstrap_srv/src/space.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs:89]: Since the infos are only stored on disk, the logic to update them means reading all the infos individually then doing any reorganization of the list of read references needed and then updating the specific reference list in memory that the read requests are cloning.

Since the read references themselves are not changing at all, it's that specific list in memory that readers are cloning that is okay for them to keep cloning. They will just get the older list (that may contain some recently expired infos, but that's okay).

Then that is locked and swapped out atomically, so thereafter read requests will get the updated list.

(no worries, understanding is what code review is all about! : )

File: crates/bootstrap_srv/src/server.rs
Comment (zippy) [inline crates/bootstrap_srv/src/server.rs:23]: Is there any value in passing the error through here?

Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:23]: What would you like to pass this through? This is a drop impl, we don't have many options. Technically you're not supposed to panic in drop impls, you'll get a stack dump that is not very useful... I'm open to other suggestions... but rust's handling of panics in other threads is pretty poor, I haven't really heard of a better pattern.

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:167]: [Created Issue](https://github.com/holochain/kitsune2/issues/17)

File: crates/bootstrap_srv/src/space.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs:8]: Sure, I can start with just a typedef for code clarity. I don't think there's any need for a newtype yet.

File: crates/bootstrap_srv/src/store.rs
Comment (zippy) [inline crates/bootstrap_srv/src/store.rs:38]: Should this also be in the config?

Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:38]: The pressure leading to wanting this larger is the inefficiency of having multiple tempfiles open. This lets us store more than 10_000 entries in one file (because infos are often much less than 1024 bytes long). It doesn't feel like we should need this to be much larger.

The pressure leading to wanting this smaller is the constraints of disk space. On a 16 core system (which cloud servers will likely not have anywhere near that) the default production config will have 64 active write tempfiles, giving 64MiB. Even if we have 10 older generations still around, that's only 640 MiB. Do we think servers will ever have that little disk space? Or to look at it from the opposite perspective, let's say the cloud server provides 10GiB of disk space. We can store 10,485,760 agent records at once, even if many of those are defunct, waiting to be cleaned up.

I don't *feel* like dev-ops folks are going to understand the consequences of making changes to this number. Providing the option to may just be confusing, and not practically useful. What do you think?

P.S. Re-reading this post, I'm not sure I even described it without being confusing...

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

File: crates/bootstrap_srv/src/space.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

File: crates/bootstrap_srv/src/space.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/space.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae

**Review by jost-s — COMMENTED** [review]
Seems great! Pruning is tested in the other PR - so just a couple of questions.

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs:33]: It needn't be the main thread if the server were passed to a different thread. But yes, whichever thread the server instance variable is dropped on will panic.

File: crates/bootstrap_srv/src/parse.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/parse.rs]: Good call, done: https://github.com/holochain/kitsune2/pull/10/commits/a4ac1e1156bc1026f9df6e19743edb0b22ad1e15

File: crates/bootstrap_srv/src/server.rs
Comment (jost-s) [reply crates/bootstrap_srv/src/server.rs:33]: Won't that stop the whole server when it's only about dropping workers?

Comment (ThetaSinner) [reply crates/bootstrap_srv/src/server.rs:33]: I'd suggest that this doesn't have to be a drop implementation here. The `main` program is doing the drop explicitly but you only need to create one instance of this per program, so it could equally be a `shutdown` method that can return an error?

Comment (ThetaSinner) [reply crates/bootstrap_srv/src/server.rs:33]: A drop implementation that calls `shutdown().unwrap()` would be good, to make sure that the user gets some attempt at shutdown if they forget to call it

File: crates/bootstrap_srv/src/server.rs
Comment (ThetaSinner) [inline crates/bootstrap_srv/src/server.rs:261]:
```suggestion
        // validate signature (do this at the end because it's more expensive)
```

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [inline crates/bootstrap_srv/src/store.rs:1]: This module is a little scary to me :) The documentation makes sense, and I can follow the code, I'm not sure how much I can asses the behavior

File: crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:48]: No, the ctrl-c hook gives control of ending the process to the process itself.

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [inline crates/bootstrap_srv/src/store.rs:134]: This function reads as though if it's called fast enough across a number of threads, it will keep having to allocate new temp files. Because each call will lock/remove and then write the new content and then lock/insert, any time the queue is empty, a new file is created. Should the number of files be bounded here? That is how I understood the documentation above

**Review by ThetaSinner — COMMENTED** [review]
Still working through the store, will pick up after meetings

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:27]: The tempfile drop implementation indeed cleans up the tempfile. If that fails for any reason (panic on the thread) then it is up to the os to clean it up. That's completely [system specific](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#operating-system-specific-concerns).

File: crates/bootstrap_srv/src/config.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/config.rs:28]: No, this is just for bootstrap. I'd prefer to handle more spaces and less agents per space than the other way around. Once they are in the space, they can get additional peers by gossip.

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:29]: Almost. We're writing the individual agent infos to the store, and keeping the space list with only the references in memory. So when we update, only new infos are written to the disk. This avoids us re-writing the same infos over and over every time the space updates.

Yes indeed, the max expires_at is 30 min past the created_at and the max in the future created_at can be is 3 min, so the max we'll hold an individual info is 33 min plus the interval the prune task runs on.

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:72]: I thought I did : ) "This will extend the provided buffer with the bytes read from the store."

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:69]: Yep, exactly.

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:27]: I'd optionally include that in the documentation here but maybe it's too much detail

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:29]: I see, so I'd slightly misunderstood what's being written. So that's more efficient than I was thinking :)

I see the logic of not holding the info for more than 33 minutes ish. However, that doesn't prevent rapid agent updates right? So I could send a new agent info every 5ms and it would force the space required by temporary files to grow rapidly. It would expire after 33 minutes so the files could be dropped but still, you could add quite a load before the first ones start expiring

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:72]: I see! Oh I get how this works now with the code in a file above with `Space::read`, for building an array from the content. Okay cool, I get it now :)

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:29]: Your logic is correct. I'd say this will be mitigated by rate limiting.

We could potentially use a different strategy such as having a tempfile per space, but that inevitably increases system resource usage (like max file handles).

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:134]: Right, that makes sense with the docs further up. Thank you

File: crates/bootstrap_srv/src/store.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/store.rs:1]: I agree. It would be nice to find a way to test it, at least a little bit. Perhaps if I made the max file size configurable at least internally, and then wrote some tests that explicitly dropped read handles and checked if the files still exist on-disk? I don't know if that would be flaky or not...

Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:1]: I've been through this in detail now and I'm much happier. Thank you for letting me try and fail to pick it apart so I could make sense of it :)

File: crates/bootstrap_srv/src/store.rs
Comment (ThetaSinner) [reply crates/bootstrap_srv/src/store.rs:29]: Cool, yes that sounds like a rate limiting issue - happy for this to be solved elsewhere :)

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/ebac6ff5dcaff4f2b78c66fab4e696e75c504847

File: crates/bootstrap_srv/src/parse.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/parse.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/85dffc79aec3d3d5493b4cab9c9ddabce0bc6bfb

File: crates/bootstrap_srv/src/server.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/server.rs]: Done: https://github.com/holochain/kitsune2/pull/10/commits/b407784ebfeca5aa999c249cce3f2871ed1350ff

**Review by mattyg — APPROVED** [review]

File: crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs
Comment (neonphog) [reply crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:19]: I'm going to decline this because if I change anything it'll dismiss the approval. But also because I don't really care about todos in the code. Feel free to make a PR to remove them! : )

Comment (jost-s) [reply crates/bootstrap_srv/src/bin/kitsune2-bootstrap-srv.rs:19]: > But also because I don't really care about todos in the code.

Exactly, no one does! :grinning:

**Pull request merged**
